### PR TITLE
Rollback memtable flush upon atomic flush fail

### PR DESCRIPTION
**Pull request by riversand963**
This fixes an assertion.

An atomic flush can have multiple flush jobs. Some of them may fail. If any of
them fails, we need to rollback all of them.
For the flush jobs that do fail, we already call `RollbackMemTableFlush` in
`FlushJob::Run`. The tricky part is for flush jobs that have completed
successfully. We need to call `RollbackMemTableFlush` for them as well.

The newly added DBAtomicFlushTest.AtomicFlushRollbackSomeJobs will SigAbort
without the corresponding change in AtomicFlushMemTablesToOutputFiles.

Test plan:
```
$make clean && make -j32 all check
$./db_flush_test
```

### CODE REVIEW

File: db/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -426,9 +438,11 @@ Status DBImpl::AtomicFlushMemTablesToOutputFiles(
     }
     if (!s.IsShutdownInProgress()) {
       for (int i = 0; i != num_cfs; ++i) {
-        auto& mems = jobs[i].GetMemTables();
-        cfds[i]->imm()->RollbackMemtableFlush(mems,
-                                              file_meta[i].fd.GetNumber());
+        if (exec_status[i].first && exec_status[i].second.ok()) {
```
Comment (riversand963) [inline db/db_impl_compaction_flush.cc]: This is the important part of this change.

File: db/db_impl_compaction_flush.cc [hunk 2]
```diff
@@ -344,6 +344,18 @@ Status DBImpl::AtomicFlushMemTablesToOutputFiles(
         break;
       }
     }
+    if (num_cfs > 1) {
+      TEST_SYNC_POINT(
+          "DBImpl::AtomicFlushMemTablesToOutputFiles:SomeFlushJobsComplete:1");
+      TEST_SYNC_POINT(
+          "DBImpl::AtomicFlushMemTablesToOutputFiles:SomeFlushJobsComplete:2");
+    }
+    exec_status[0].second =
```
Comment (ajkr) [inline db/db_impl_compaction_flush.cc]: do we need to do this even when `s` is non-ok?

also it's not so obvious to me why we run jobs in the order 1, ..., `num_cfs-1`, 0.

Comment (riversand963) [reply db/db_impl_compaction_flush.cc]: > do we need to do this even when `s` is non-ok?
What are you referring to by 'this'?
>  also it's not so obvious to me why we run jobs in the order
In future, if we want to parallelize the flush jobs, then this current thread runs job[0], while other num_cfs-1 threads run jobs 1, ...num_cfs-1

Comment (riversand963) [reply db/db_impl_compaction_flush.cc]: Discussed offline, we should check `s` before running `jobs[0]`, at least for now.

**Comment by riversand963** [comment]
Thanks @ajkr for the feedback!

**Pull request closed without merging**